### ISSUE TITLE
Conda 4.3 cannot be used on Windows. Unlinking seems not to happen, post install scripts are not run.

### ISSUE THREAD

**Issue by mingwandroid** [conda/conda#4105]
I believe there are likely two different bugs in this report (failure to unlink, failure to run post install).

In an installation with qt already installed:
```
conda list --show-channel-urls
qt                        5.6.2                    vc14_0  [vc14]  defaults
```
..  `conda install qt --force` does not manage to complete (it seems conda does not unlink and also bails out before the post install script gets run). Here are some of most relevant looking parts of the `--debug` output.

```
DEBUG conda.instructions:execute_instructions(243): executing plan
[('PREFIX', 'C:\\Users\\ray\\mc-x64-3.5'), ('PRINT', 'Linking packages ...'),
 ('PROGRESSIVEFETCHEXTRACT', <conda.core.package_cache.ProgressiveFetchExtract object at 0x000001373D039A58>),
 ('UNLINKLINKTRANSACTION', ((), (Dist(_Dist__initd=True, channel='defaults', dist_name='qt-5.6.2-vc14_0', with_features_depends=None, base_url=None, platform=None),))),
 ('LINK', Dist(_Dist__initd=True, channel='defaults', dist_name='qt-5.6.2-vc14_0', with_features_depends=None, base_url=None, platform=None)),
 ('SYMLINK_CONDA', 'C:\\Users\\ray\\mc-x64-3.5')]




DEBUG conda.instructions:execute_instructions(249):  UNLINKLINKTRANSACTION(((), (Dist(_Dist__initd=True, channel='defaults', dist_name='qt-5.6.2-vc14_0', with_features_depends=None, base_url=None, platform=None),)))
DEBUG conda.core.link:create_from_dists(114): instantiating UnlinkLinkTransaction with
  target_prefix: C:\Users\ray\mc-x64-3.5
  unlink_dists:

  link_dists:
    defaults::qt-5.6.2-vc14_0




, CondaVerificationError: The package 'defaults::qt-5.6.2-vc14_0' cannot be installed due to a
path collision for 'library/bin/qt5clucene.dll'.
This path already exists in the target prefix, and it won't be removed
by an uninstall action in this transaction. The path is one that conda
doesn't recognize. It may have been created by another package manager.
If you'd like to proceed anyway, re-run the command with the `--force`
flag.
(this is repeated for every file).

INFO conda.gateways.disk.create:create_link(220): file exists, but clobbering: 'C:\\Users\\ray\\mc-x64-3.5\\Library\\include\\qt\\QtCore\\qfuture.h'
(as is this)




.. then finally ..
INFO conda.gateways.disk.create:create_link(220): file exists, but clobbering: 'C:\\Users\\ray\\mc-x64-3.5\\Scripts\\.qt-post-link.bat'
DEBUG conda.instructions:execute_instructions(249):  LINK(Dist(_Dist__initd=True, channel='defaults', dist_name='qt-5.6.2-vc14_0', with_features_depends=None, base_url=None, platform=None))
DEBUG conda.instructions:execute_instructions(249):  SYMLINK_CONDA('C:\\Users\\ray\\mc-x64-3.5')
DEBUG conda.instructions:SYMLINK_CONDA_CMD(104): Conda environment at C:\Users\ray\mc-x64-3.5 start with '_'. Skipping symlinking conda.

```

Observations:
1. The filenames in the path collision are all lowercase (library/bin/qt5clucene.dll).
2. The `UNLINKLINKTRANSACTION` has no `unlink_dists`.
3. Despite force and the installation proceeding and a `.qt-post-link.bat` getting written, it is not executed.

**Comment by kalefranz** [comment]
I'll try to get to this today.  I'm about to tag `4.3.1`, so the fix will have to be released in the next round.  FWIW @mingwandroid, in this situation looking at logs, the `-vvv` flag might be helpful. `-v` is INFO level, `-vv` is DEBUG level, and `-vvv` is TRACE level.

**Comment by kalefranz** [comment]
Can you give me whatever commands I need to use to exactly reproduce the problem?  Like create an environment with so-and-so specs, activate the environment, then run so-and-so command...

**Comment by kalefranz** [comment]
@mingwandroid 

> The filenames in the path collision are all lowercase (library/bin/qt5clucene.dll).

I fixed this in 4.3.1.  https://github.com/conda/conda/blob/4.3.1/conda/core/link.py#L196-L201

**Comment by kalefranz** [comment]
> The UNLINKLINKTRANSACTION has no unlink_dists.

I can confirm reproducing this.

**Comment by kalefranz** [comment]
> Despite force and the installation proceeding and a .qt-post-link.bat getting written, it is not executed.

I see an obvious bug here.

**Comment by mingwandroid** [comment]
> Can you give me whatever commands I need to use to exactly reproduce the problem?

Sorry I was out visiting relatives.

> I fixed this in 4.3.1. https://github.com/conda/conda/blob/4.3.1/conda/core/link.py#L196-L201

By default, Windows is `case-insensitive` but `case-preserving`, however it is entirely possible to run Windows case-sensitively (I've run it this way before with no problems): http://superuser.com/a/842670. Regardless, I'm not sure why anything's being made lowercase here on Windows, can you explain the reason for that?

macOS is, by default the exact same (though here you can format as case sensitive instead of being able to toggle a setting), so the logic for making things lower-case in that code should apply to macOS too I guess (but only after checking that registry key on Windows and the filesystem type on macOS first).

**Comment by kalefranz** [comment]
I struggled with the window pre-check code.

>  Regardless, I'm not sure why anything's being made lowercase here on Windows, can you explain the reason for that?

Check out and compare the win-64 `python-3.4.5-0` and `python-3.5.2-0` packages.  Compare, for example, `Tools/Scripts/2to3.py` to `Tools/scripts/2to3.py`.

#pain

**Comment by kalefranz** [comment]
I guess I only ran into the case-insensitivity problems on windows so far.  Maybe we need to treat macos just like we're treating windows here?

**Comment by kalefranz** [comment]
For the pre/post link/unlink part: https://github.com/conda/conda/pull/4113

**Comment by kalefranz** [comment]
@mingwandroid know of any packages lighter than qt that I can use in an integration test for post link scripts?

**Comment by kalefranz** [comment]
nevermind I got it I think

**Comment by kalefranz** [comment]
I think contributing to the problem here is that we've overloaded the meaning of `--force`, and we need to disentangle those.

**Comment by kalefranz** [comment]
This should now be resolved with #4113 and subsequent PRs.
